### ISSUE TITLE
Recursive union type silently breaks union structure

### ISSUE THREAD

**Issue by deiteris** [aml-org/amf#1773]
Your issue may already be reported! Please search on Github issues before creating one.

* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] question
  - [ ] other

* **What is the current behavior?**

When a property of the union type refers to the same union type, this breaks the union structure by nesting them into each other. The same thing happens when a non-existent type is inserted into the union.

* **If the current behavior is a bug, please provide the minimal steps to reproduce with AMF code. Please don't upload AMF code mixed with yours.**

The following code with RAML reproduces the issue:
```js
import amf from 'amf-client-js'

const raml = `
#%RAML 1.0 Library

types:
  TypeOne:
    properties:
      prop3: string | integer

  TypeTwo:
    properties:
      prop1: string
      prop2?: string

  TypeThree:
    properties:
      prop5: boolean

  TypeMap:
    properties:
      prop4: Value

  Value:
    type: TypeOne | TypeTwo | TypeThree | TypeMap # Broken when TypeMap is present
`;

(async () => {
  const client = amf.RAMLConfiguration.RAML10().baseUnitClient();

  const data = await client.parseContent(raml);

  const validation = await client.validate(data.baseUnit.cloneUnit());
  const report = data.merge(validation);
  console.log(report.toString())

  function traverse(type) {
    console.log(type.id)
    for (const parent of type.inherits) {
      traverse(parent)
    }
    if ('anyOf' in type) {
      for (const t of type.anyOf) {
        traverse(t)
      }
    }
  }
  traverse(data.baseUnit.declares.find(shape => shape.name.value() === 'Value'))
})();
```

The output:
```
Model: http://a.ml/amf/default_document
Conforms: true
Number of results: 0

http://a.ml/amf/default_document#/declares/shape/Value
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/shape/default-node
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/union/default-union
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/union/default-union/anyOf/shape/default-node
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/union/default-union/anyOf/union/default-union
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/union/default-union/anyOf/union/default-union/anyOf/shape/default-node
http://a.ml/amf/default_document#/declares/shape/Value/inherits/union/default-union/anyOf/union/default-union/anyOf/union/default-union/anyOf/shape/default-node_2
```

* **What is the expected behavior?**

`inherits[0].anyOf` of the `Value` type must contain 4 shapes instead of them getting nested into each other.

* **What is the motivation/use case for changing the behavior? (for feature requests)**

* **Please tell us about your environment:**
  
  - AMF Version: 5.2.6
  - AMF Distribution: JS
  - Operating System: Windows 10 22H2

**Comment by looseale** [comment]
Hi @deiteris 
We are already working on a refactor in the transformation code that resolves inheritance because we also detected some cases of inheritance mixed with recursion that have this same problem.
We still don't have an ETA for this, but it probably will be in the late May release.

**Comment by tomsfernandez** [comment]
Hi @deiteris! This refactor was released in version 5.3.0. The latest version has some fixes on that inheritance resolution also. Do you want to test it to see if this still applies?

**Comment by deiteris** [comment]
Hi @tomsfernandez! The code I mentioned in the issue produces identical result in 5.3.0-0 and 5.4.0-1.

I've tried different transformation pipelines as well which also produce the same result (even though a more readable one). For example, this is the result of the Default pipeline:
```
http://a.ml/amf/default_document#/declares/union/Value
http://a.ml/amf/default_document#/declares/shape/TypeOne
http://a.ml/amf/default_document#/declares/union/Value/inherits/union/default-union/anyOf/union/default-union
http://a.ml/amf/default_document#/declares/shape/TypeTwo
http://a.ml/amf/default_document#/declares/union/Value/inherits/union/default-union/anyOf/union/default-union/anyOf/union/default-union   
http://a.ml/amf/default_document#/declares/scalar/TypeThree
http://a.ml/amf/default_document#/declares/shape/TypeMap
```

What I also find concerning, is that this nesting seems to happen with any custom type even without recursive type (I guess might need to rename the issue). So, scalar `TypeOne | TypeTwo | TypeThree` (that are just `string | number | integer`) will also produce this sequence with nesting:
```
http://a.ml/amf/default_document#/declares/union/Value
http://a.ml/amf/default_document#/declares/scalar/TypeOne
http://a.ml/amf/default_document#/declares/union/Value/inherits/union/default-union/anyOf/union/default-union
http://a.ml/amf/default_document#/declares/scalar/TypeTwo
http://a.ml/amf/default_document#/declares/scalar/TypeThree
```

While union of plain `string | number | integer` produces correct sequence:
```
http://a.ml/amf/default_document#/declares/union/PlainUnion
http://a.ml/amf/default_document#/declares/union/PlainUnion/inherits/union/default-union/anyOf/scalar/default-scalar
http://a.ml/amf/default_document#/declares/union/PlainUnion/inherits/union/default-union/anyOf/scalar/default-scalar_1
http://a.ml/amf/default_document#/declares/union/PlainUnion/inherits/union/default-union/anyOf/scalar/default-scalar_2
```

I've been making my own expression parser and it seems to me as if CST which was used to generate AST is as follows:
```
(Value: union 
  (null: union (TypeOne: custom_type)  (TypeTwo: custom_type)) 
  (TypeThree: custom_type)
)
```
Which may happen in case `|` is treated as an operator and associated to left/right value.

While what I expect (and managed to achieve) is:
```
(Value: union (TypeOne: custom_type) (TypeTwo: custom_type) (TypeThree: custom_type)))
```

**Comment by deiteris** [comment]
As a workaround, I've flatten such union structure in my own code (v5.4.0-1) in the `flattenUnion` function, just in case someone finds it useful:
```js
import amf, { Module, Shape, TypeIRI, TypeUtil, UnionShape } from 'amf-client-js'

const raml = `
#%RAML 1.0 Library

types:
  TypeOne:
    properties:
      prop3: string | integer

  TypeTwo:
    properties:
      prop1: string
      prop2?: string

  TypeThree:
    properties:
      prop5: boolean

  TypeMap:
    properties:
      prop4: Value

  Value:
    type: TypeOne | TypeTwo | TypeThree | TypeMap
`;

(async () => {
  const client = amf.RAMLConfiguration.RAML10().baseUnitClient()

  const data = await client.parseContent(raml)

  //const result = client.transform(data.baseUnit, amf.PipelineId.Default)

  function flattenUnion(shape: UnionShape, shapes: Shape[] = []): Shape[] {
    for (const t of shape.anyOf) {
      if (TypeUtil.isTypeOf(t, TypeIRI.UnionShape) && t.name.isNull) {
        flattenUnion(t as UnionShape, shapes)
      } else {
        shapes.push(t)
      }
    }
    return shapes
  }

  function traverse(type: Shape) {
    console.log(type.isLink ? type.linkLabel.value() : type.name.value())
    for (const parent of type.inherits) {
      traverse(parent)
    }
    if (TypeUtil.isTypeOf(type, TypeIRI.UnionShape)) {
      const shapes = flattenUnion(type as UnionShape)
      for (const t of shapes) {
        traverse(t)
      }
    }
  }
  traverse(((data.baseUnit as Module).declares as Shape[]).find(shape => shape.name.value() === 'Value')!)
})();
```

Outputs:
```
Value
null
TypeOne
TypeTwo
TypeThree
TypeMap
```

Without this workaround, the output is:
```
Value
null
TypeOne
null
TypeTwo
null
TypeThree
TypeMap
```

**Comment by deiteris** [comment]
Looks like it's fixed with 5.4.1.

**Comment by gutee** [comment]
Thanks! Closing the issue.
